### PR TITLE
reactive-dict extended api

### PR DESCRIPTION
**Pull request by awatson1978**
Pull request for #3574.

### CODE REVIEW

File: packages/reactive-dict/reactive-dict.js [hunk 1]
```diff
@@ -124,6 +139,47 @@ _.extend(ReactiveDict.prototype, {
...
+    // delete Session.keys['foo']
+
+    // might be better implemented along the lines of
+    // delete self.keyValueDeps[key][serializedValue];
+
+    self.set(key, undefined);
+  },
+  _setObject: function(object){
+    var self = this;
+    for (var key in object) {
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: This can be simplified with _.each: http://underscorejs.org/#each

File: packages/reactive-dict/reactive-dict.js [hunk 2]
```diff
@@ -124,6 +139,47 @@ _.extend(ReactiveDict.prototype, {
...
+  toggle: function(key){
+    var self = this;
+
+    // toggle currently only works on boolean values
+    // an elseif is used without a terminating else so that we can
+    // explicitely handle the true and false cases
+    // while leaving null and undefined cases alone
+    if(self.get(key) === true){
+      self.set(key, false);
+    }else if(self.get(key) === false){
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: Please follow our code style guidelines regarding spaces around language keywords. (Always put spaces) https://github.com/meteor/meteor/wiki/Meteor-Style-Guide

File: packages/reactive-dict/reactive-dict.js [hunk 3]
```diff
@@ -36,25 +36,40 @@ ReactiveDict = function (dictName) {
...
-    changed(self.keyDeps[key]);
-    if (self.keyValueDeps[key]) {
-      changed(self.keyValueDeps[key][oldSerializedValue]);
-      changed(self.keyValueDeps[key][value]);
+    // if the first argument is an object and the second is undefined
+    // we treat it as an object of key/value pairs, and iterate through it
+    } else if ((typeof key_or_object === 'object') && (value === undefined)) {
+      self._setObject(key_or_object);
+    } else {
+      throw new Error("Key must be a string or object of key/value pairs: " + key);
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: I think this would be better if it said "First argument to set must be ...."

File: packages/reactive-dict/reactive-dict.js [hunk 4]
```diff
@@ -36,25 +36,40 @@ ReactiveDict = function (dictName) {
...
+  // set() began as a key/value method, but we are now overloading it
+  // to take an object of key/value pairs, similar to backbone
+  // http://backbonejs.org/#Model-set
+
+  set: function (key_or_object, value) {
     var self = this;
 
-    value = stringify(value);
+    // if the first argument is a string, use the original key/value method
+    if (typeof key_or_object === 'string') {
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: To avoid wrapping this whole block in an if, let's just have a short-circuit return in the front:

```
if ((typeof key_or_object === 'object') && (value === undefined)) {
   self._setObject(key_or_object);
   return;
}

// Then the rest of the code
```

File: packages/reactive-dict/reactive-dict.js (hunk 3 above)
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: Instead of having this in the else, let's put the argument check at the top of the function.

Comment (awatson1978) [inline packages/reactive-dict/reactive-dict.js]: Sounds like a plan!

File: packages/reactive-dict/reactive-dict.js (hunk 2 above)
Comment (awatson1978) [inline packages/reactive-dict/reactive-dict.js]: Consider it done.

File: packages/reactive-dict/reactive-dict.js [hunk 5]
```diff
@@ -124,6 +139,47 @@ _.extend(ReactiveDict.prototype, {
...
   },
+  clear: function(key){
+    var self = this;
+    self.set(key, null);
+    return true;
+  },
+  toggle: function(key){
+    var self = this;
+
+    // toggle currently only works on boolean values
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: I think there's a decision to be made here. IMO it would make sense for toggle to work just like `Session.set("x", ! Session.get("x"))`, since that's the pattern we are trying to replace?

File: packages/reactive-dict/reactive-dict.js (hunk 1 above)
Comment (awatson1978) [inline packages/reactive-dict/reactive-dict.js]: Will do.

File: packages/reactive-dict/reactive-dict.js (hunk 5 above)
Comment (awatson1978) [inline packages/reactive-dict/reactive-dict.js]: Um...  my experience has been that there are a lot of edge cases around null, undefined, and empty strings.  It's not clear how ``!null`` or ``!undefined`` or ``!""`` should be handled.  What's the negative of a object?  Personally, I never use ``!Session.get("x")`` in my code, precisely because of those edge cases.  So it's not exactly the pattern I was trying to replace when I created this.

**Comment by awatson1978** [comment]
Ah.  Tweaking my prettify script a bit more to match meteor styleguide.

File: packages/reactive-dict/reactive-dict.js (hunk 5 above)
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: How do you usually implement toggle in your apps?

File: packages/reactive-dict/reactive-dict.js [hunk 6]
```diff
@@ -36,25 +36,34 @@ ReactiveDict = function (dictName) {
...
+    if ((typeof key_or_object === 'object') && (value === undefined)) {
+      self._setObject(key_or_object);
+      return;
+    }
+
     value = stringify(value);
 
     var oldSerializedValue = 'undefined';
-    if (_.has(self.keys, key)) oldSerializedValue = self.keys[key];
+    if (_.has(self.keys, key_or_object)) oldSerializedValue = self.keys[key_or_object];
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: I would add a `var key = key_or_object` once we have established we are dealing with a string key, so that the code below can remain unchanged.

File: packages/reactive-dict/reactive-dict.js [hunk 7]
```diff
@@ -94,19 +103,19 @@ _.extend(ReactiveDict.prototype, {
     // XXX we could allow arrays as long as we recursively check that there
     // are no objects
     if (typeof value !== 'string' &&
-        typeof value !== 'number' &&
-        typeof value !== 'boolean' &&
-        typeof value !== 'undefined' &&
-        !(value instanceof Date) &&
-        !(ObjectID && value instanceof ObjectID) &&
-        value !== null)
+      typeof value !== 'number' &&
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: Please don't change indentation unnecessarily.

File: packages/reactive-dict/reactive-dict.js [hunk 8]
```diff
@@ -94,19 +103,19 @@ _.extend(ReactiveDict.prototype, {
...
+      !(ObjectID && value instanceof ObjectID) &&
+      value !== null)
       throw new Error("ReactiveDict.equals: value must be scalar");
     var serializedValue = stringify(value);
 
     if (Tracker.active) {
       self._ensureKey(key);
 
-      if (! _.has(self.keyValueDeps[key], serializedValue))
+      if (!_.has(self.keyValueDeps[key], serializedValue))
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: Space after !

File: packages/reactive-dict/reactive-dict.js [hunk 9]
```diff
@@ -114,7 +123,7 @@ _.extend(ReactiveDict.prototype, {
         Tracker.onInvalidate(function () {
           // clean up [key][serializedValue] if it's now empty, so we don't
           // use O(n) memory for n = values seen ever
-          if (! self.keyValueDeps[key][serializedValue].hasDependents())
+          if (!self.keyValueDeps[key][serializedValue].hasDependents())
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: Please don't run a "prettifier" over the whole file when submitting a pull request, it results in a lot of unnecessary changes.

File: packages/reactive-dict/reactive-dict.js (hunk 7 above)
Comment (awatson1978) [inline packages/reactive-dict/reactive-dict.js]: Sorry.  Before I do another pull-request, I'll sit down and give my prettify script a close audit and make sure that it's in alignment with Meteor styleguides.  I ran it to add spaces around language keywords, and it wound up being one step forward and one step backwards.

File: packages/reactive-dict/reactive-dict.js [hunk 10]
```diff
@@ -124,6 +133,48 @@ _.extend(ReactiveDict.prototype, {
...
+    // while leaving null and undefined cases alone
+    if (self.get(key) === true) {
+      self.set(key, false);
+    } else if (self.get(key) === false) {
+      self.set(key, true);
+    }
+    return true;
+  },
+
+  _remove: function (key) {
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: Let's not add this method right now, especially because there are no tests and it will be unclear how it differs from `clear`

File: packages/reactive-dict/reactive-dict.js (hunk 9 above)
Comment (awatson1978) [inline packages/reactive-dict/reactive-dict.js]: Yeah, my bad.  Hotkey reflex.  I'll get that sorted out.

File: packages/reactive-dict/reactive-dict.js (hunk 10 above)
Comment (awatson1978) [inline packages/reactive-dict/reactive-dict.js]: Yup.

File: packages/reactive-dict/reactive-dict.js (hunk 5 above)
Comment (awatson1978) [inline packages/reactive-dict/reactive-dict.js]: I tend to be boring and explicit and avoid one-liners because of edge cases.  Usually just a variation on the following:
````js
if(Session.get('foo')){  
  Session.set('foo',false);     
}else{
  Session.set('foo',true);
}
````
I

Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: I'm pretty sure this ends up with the exact same result as `! Session.get('foo')`

File: packages/reactive-dict/reactive-dict.js (hunk 7 above)
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: I'm saying even if parts of the file don't meet the style guide, we shouldn't change them as part of this PR, in which case no prettifier will do the right thing.

Comment (awatson1978) [inline packages/reactive-dict/reactive-dict.js]: Understood.  No styleguide scripts.  Hand editing.  I just disabled it from my editor while I do any more edits.

**Comment by stubailo** [comment]
I think from some of the discussions here it's become clear that some people might want different things from `clear` or `toggle` (setting to `undefined` vs `null`, accepting non `true` or `false` values).

For new, let's just leave the original plan of having `set` accept an object and merge that in, then talk about the other features some other time.

File: packages/reactive-dict/reactive-dict.js [hunk 11]
```diff
@@ -36,9 +36,21 @@ ReactiveDict = function (dictName) {
 };
 
 _.extend(ReactiveDict.prototype, {
-  set: function (key, value) {
+  // set() began as a key/value method, but we are now overloading it
+  // to take an object of key/value pairs, similar to backbone
+  // http://backbonejs.org/#Model-set
+
+  set: function (key_or_object, value) {
```
Comment (stubailo) [inline packages/reactive-dict/reactive-dict.js]: Meteor style guide says camel case for variables: `keyOrObject`.

**Comment by awatson1978** [comment]
How about we just put them as internal methods ``_toggle()`` and ``_clear()``?  There's useful functionality there, even if it's not entirely clear about the edge cases.

**Comment by stubailo** [comment]
I think they could easily be in a package as well.

**Comment by awatson1978** [comment]
Fine.  Whatever.  Sorry I tried to contribute.  I got other work I have to get back to.

**Comment by svanderbleek** [comment]
Toggle should be value preserving unless `true` or `false`, it's pretty simple. This should have gone through and it's confusing why not. Clear should just set to whatever a value is before it is set, but it's not as useful.

**Comment by stubailo** [comment]
Merged in this commit: 26fc3d6666d78ed99f71e1f2e2a6216782b60546

I'm sorry it got complicated.
